homeassistant/components/songpal/coordinator.py

```diff
@@ -0,0 +1,264 @@
+    initialized: bool = False
+
+    def __init__(
+        self, hass: HomeAssistant, config_entry: ConfigEntry, name: str, device: Device
+    ) -> None:
+        """Initialize coordinator."""
+
+        super().__init__(
+            hass,
+            _LOGGER,
+            name=f"{DOMAIN} ({config_entry.unique_id})",
+            update_method=self.update,
+            update_interval=timedelta(seconds=10),
+        )
+
+        self.device_name = name
+        self.data = {}
+        self.device = device
+        self._initialized = False
+
+    async def _async_setup(self):
+        await self.async_activate_websocket()
+        await self.full_refresh()
+        self.async_set_updated_data(self.data)
+
+    async def update(
+        self,
+        log_failures=True,
+        raise_on_auth_failed=False,
+        scheduled=False,
+        raise_on_entry_error=False,
+    ):
+        """Poll for data updates that are not pushed."""
+        self.data = await self.polling_refresh(self.data)
+        return self.data
+
+    async def async_shutdown(self) -> None:
+        """Run when entity will be removed from hass."""
+        await self.device.stop_listen_notifications()
+        await super().async_shutdown()
+
+    async def async_activate_websocket(self):
+        """Activate websocket for listening if wanted."""
+        _LOGGER.warning("Activating websocket connection")
+
+        async def _volume_changed(volume: VolumeChange):
+            _LOGGER.debug("Volume changed: %s", volume)
+
+            self.data["volumes"] = await self.device.get_volume_information()
+            self.async_set_updated_data(self.data)
+
+        async def _source_changed(content: ContentChange):
+            _LOGGER.debug("Source changed: %s", content)
+
+            self.data["inputs"] = await self.device.get_inputs()
+            self.async_set_updated_data(self.data)
+
+        async def _setting_changed(setting: SettingChange):
+            _LOGGER.debug("Setting changed: %s", setting)
+
+            data = self.data
+            data["sound_settings"] = await self.device.get_sound_settings()
+            data["bluetooth_settings"] = await self.device.get_bluetooth_settings()
+            data["misc_settings"] = await self.device.get_misc_settings()
+            data["playback_settings"] = await self.device.get_playback_settings()
+            self.data = data
+            self.async_set_updated_data(self.data)
+
+        async def _power_changed(power: PowerChange):
+            _LOGGER.debug("Power changed: %s", power)
+
+            self.data["power"] = await self.device.get_power()
+            self.async_set_updated_data(self.data)
+
+        async def _try_reconnect(connect: ConnectChange):
+            _LOGGER.warning(
+                "[%s(%s)] Got disconnected, trying to reconnect",
+                self.name,
+                self.device.endpoint,
+            )
+            _LOGGER.debug("Disconnected: %s", connect.exception)
```

Comment:
Maybe the visible warning should be extended to contain the exception string, and then this could be removed altogether?